modules/sasl.cpp

```diff
@@ -24,9 +32,132 @@ static const struct {
 } SupportedMechanisms[] = {
 	{ "EXTERNAL",           "TLS certificate, for use with the *cert module", false },
 	{ "PLAIN",              "Plain text negotiation, this should work always if the network supports SASL", true },
+#ifdef HAVE_LIBSSL
+	{ "ECDSA-NIST256P-CHALLENGE", "ECDSA public key authentication.", true },
+#endif
 	{ nullptr, nullptr, false }
 };
 
+#ifdef HAVE_LIBSSL
+class ECDACommon {
+private:
+	EC_KEY *key;
+	bool loaded;
+
+public:
+	ECDACommon() {
+		key = EC_KEY_new_by_curve_name(NID_X9_62_prime256v1);
+		EC_KEY_set_conv_form(key, POINT_CONVERSION_COMPRESSED);
+		loaded = false;
+	}
+
+	~ECDACommon() {
+		if (key)
+		    EC_KEY_free(key);
+	}
+
+	/* Generate an ECC keypair suitable for authentication. */
+	void GenerateKey(void) {
+		EC_KEY_generate_key(key);
+		loaded = true;
+	}
+
+	/* Get base64 encoded private key from key */
+	CString GetPubBase64(void) {
+		CString ret;
+		size_t keybuflen;
+		unsigned char *keybuf;
+
+		if (!loaded)
+		    return ret;
+
+		keybuflen = (size_t) i2o_ECPublicKey(key, NULL);
```

Comment:
`size_t uKeyBufLen = ...`
No need to list all the variables in beginning of function.